horizon/startup/offline_mode.py

```diff
@@ -0,0 +1,95 @@
+from typing import Optional, Tuple
+
+from cryptography.hazmat.primitives.kdf.hkdf import HKDF
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.backends import default_backend
+from cryptography.fernet import Fernet
+from pydantic import ValidationError
+import base64
+import secrets
+
+from horizon.startup.schemas import RemoteConfig, RemoteConfigBackup
+from opal_common.logger import logger
+
+
+class OfflineModeManager:
+    """
+    A backup for the remote config, in case the sidecar can't fetch the remote config.
+    """
+
+    def __init__(self, backup_path: str, api_key: str):
+        self._backup_path: str = backup_path
+        self._api_key = api_key
+
+    def _derive_backup_key(self, salt: Optional[bytes] = None) -> Tuple[bytes, bytes]:
+        if salt is None:
+            salt = secrets.token_bytes(16)
+        else:
+            salt = base64.urlsafe_b64decode(salt)
+
+        hkdf = HKDF(
+            algorithm=hashes.SHA256(),
+            length=32,
+            salt=salt,
+            info=b"Sidecar's local remote-config backup Key",
+            backend=default_backend(),
+        )
+        # We don't bother extracting the actual cryptographic bytes from the API key (which has a urlsafe encoding + a prefix),
+        # The 512-bit entropy is still there, and HKDF's extract phase handles inputs of non-uniform randomness.
+        key_bytes = hkdf.derive(self._api_key.encode("utf-8"))
+        return base64.urlsafe_b64encode(key_bytes), base64.urlsafe_b64encode(salt)
+
+    def backup_config(self, remote_config: RemoteConfig):
+        logger.info(
+            "Backing up remote config to {path}",
+            path=self._backup_path,
+        )
+
+        enc_key, salt = self._derive_backup_key()
```

Comment:
what about encrypting it using the api key ? wouldn't it be good to make it encrypted ? 
( would it work out of the box if you'll send the api key to this function ? )

Reply:
That was my aim to begin with. But the API key token has 64 bytes, and `Fernet` expects a 32 bytes encryption key. So I perform a standard key derivation, and adding the salt make it more secure for attacks if an attacker puts his hands on an encrypted file.

Reply:
Also - this method (deriving an encryption key from the API key using HKDF) allows the PDP to be decoupled of the way the API key is prefixed/suffixed/encoded - as long as it has 32 or more bytes of entropy - I just throw it into HKDF and get a suitable key.